beacon-chain/db/kv/attestations.go

```diff
@@ -106,81 +135,67 @@ func (k *Store) SaveAttestations(ctx context.Context, atts []*ethpb.Attestation)
-
-	buf = append(buf, []byte("root")...)
-	attRoot, err := ssz.HashTreeRoot(att)
-	if err != nil {
-		return nil, err
+// createAttestationIndicesFromData creates a set of byte indices
+// for key lookups in the DB using attestation data.
+func createAttestationIndicesFromData(attData *ethpb.AttestationData) [][]byte {
+	return [][]byte{
+		append(shardIdx, uint64ToBytes(attData.Crosslink.Shard)...),
+		append(parentRootIdx, attData.Crosslink.ParentRoot...),
+		append(startEpochIdx, uint64ToBytes(attData.Crosslink.StartEpoch)...),
+		append(endEpochIdx, uint64ToBytes(attData.Crosslink.EndEpoch)...),
 	}
-	buf = append(buf, attRoot[:]...)
-	return buf, nil
 }
 
-// ensureAttestationFilterCriteria uses a set of specified filters
-// to ensure the byte key used for db lookups contains the correct values
-// requested by the filter. For example, if a key looks like:
-// root-0x23923-parent-root-0x49349-start-epoch-3-end-epoch-4-shard-5
-// and our filter criteria wants the key to contain shard 5 and
-// start epoch 5, the key will NOT meet all the filter criteria and this
-// function will return false.
-func ensureAttestationFilterCriteria(key []byte, f *filters.QueryFilter) bool {
-	numCriteriaMet := 0
+// createAttestationIndicesFromFilters takes in filter criteria and returns
+// a list of of byte keys used to retrieve the values stored
+// for the indices from the DB.
+//
+// For attestations, these are list of hash tree roots of attestation.Data
+// objects. If a certain filter criterion does not apply to
+// attestations, an appropriate error is returned.
+func createAttestationIndicesFromFilters(f *filters.QueryFilter) ([][]byte, error) {
+	keys := make([][]byte, 0)
 	for k, v := range f.Filters() {
 		switch k {
-		case filters.Root:
-			root := v.([]byte)
-			if bytes.Contains(key, append([]byte("root"), root[:]...)) {
-				numCriteriaMet++
-			}
+		case filters.Shard:
+			idx := append(shardIdx, uint64ToBytes(v.(uint64))...)
+			keys = append(keys, idx)
 		case filters.ParentRoot:
-			root := v.([]byte)
-			if bytes.Contains(key, append([]byte("parent-root"), root[:]...)) {
-				numCriteriaMet++
-			}
+			parentRoot := v.([]byte)
+			idx := append(parentRootIdx, parentRoot...)
+			keys = append(keys, idx)
 		case filters.StartEpoch:
-			if bytes.Contains(key, append([]byte("start-epoch"), uint64ToBytes(v.(uint64))...)) {
-				numCriteriaMet++
-			}
+			idx := append(startEpochIdx, uint64ToBytes(v.(uint64))...)
+			keys = append(keys, idx)
 		case filters.EndEpoch:
-			if bytes.Contains(key, append([]byte("end-epoch"), uint64ToBytes(v.(uint64))...)) {
-				numCriteriaMet++
-			}
-		case filters.Shard:
-			if bytes.Contains(key, append([]byte("shard"), uint64ToBytes(v.(uint64))...)) {
-				numCriteriaMet++
-			}
+			idx := append(endEpochIdx, uint64ToBytes(v.(uint64))...)
+			keys = append(keys, idx)
+		default:
+			return nil, fmt.Errorf("filter criterion %v not supported for attestations", k)
```

Comment:
We directly specify the allowable criteria for filtering attestations. We wouldn't want a user trying to call `db.Attestations(ctx, filters.NewFilter().SetValidatorIndex(5))` when attestations do not have a validator index field for filtering. This is a good compromise that encourages good use of the library and prevents nasty errors/panics.